warp10/src/main/java/io/warp10/script/aggregator/Percentile.java

```diff
@@ -138,73 +163,199 @@ public int compare(Integer idx1, Integer idx2) {
+      return new Object[] {ticks[indices[j]], locations[indices[j]], elevations[indices[j]], ((Number) values[indices[j]]).doubleValue()};
+    }
 
-        if (GeoTimeSerie.NO_LOCATION != elevations[indices[m]] && GeoTimeSerie.NO_ELEVATION != elevations[indices[n]]) {
-          elevation = (long) (elevations[indices[m]] + factor * (elevations[indices[n]] - elevations[indices[m]]));
-        }
+    if (Math.abs(gamma - 1) < EPSILON) {
+      // gamma ~= 1, in that case, do not interpolate and return value at j+1
+      return new Object[] {ticks[indices[j + 1]], locations[indices[j + 1]], elevations[indices[j + 1]], ((Number) values[indices[j + 1]]).doubleValue()};
+    }
+
+    // Check values are numeric
+    if (!(values[0] instanceof Number)) {
+      throw new WarpScriptException(getName() + " can only interpolate on numeric values");
+    }
+
+    // Tick
+    long tick = Math.round((1 - gamma) * ticks[indices[j]] + gamma * ticks[indices[j + 1]]);
 
-        return new Object[] {tick, location, elevation, v};
+    // Location
+    long location;
+    if (GeoTimeSerie.NO_LOCATION == locations[indices[j]]) {
+      if (GeoTimeSerie.NO_LOCATION == locations[indices[j + 1]]) {
+        location = GeoTimeSerie.NO_LOCATION;
       } else {
-        throw new WarpScriptException("Twilight zone!");
+        location = locations[indices[j + 1]];
+      }
+    } else {
+      if (GeoTimeSerie.NO_LOCATION == locations[indices[j + 1]]) {
+        location = locations[indices[j]];
+      } else {
+        // Both locations are valid, interpolate.
+        double[] latlonj = GeoXPLib.fromGeoXPPoint(locations[indices[j]]);
+        double[] latlonjp1 = GeoXPLib.fromGeoXPPoint(locations[indices[j + 1]]);
+
+        double lat = (1 - gamma) * latlonj[0] + gamma * latlonjp1[0];
+        double lon = (1 - gamma) * latlonj[1] + gamma * latlonjp1[1];
+
+        location = GeoXPLib.toGeoXPPoint(lat, lon);
       }
     }
+
+    // Elevation
+    long elevation;
+    if (GeoTimeSerie.NO_ELEVATION == elevations[indices[j]]) {
+      if (GeoTimeSerie.NO_ELEVATION == elevations[indices[j + 1]]) {
+        elevation = GeoTimeSerie.NO_ELEVATION;
+      } else {
+        elevation = elevations[indices[j + 1]];
+      }
+    } else {
+      if (GeoTimeSerie.NO_ELEVATION == elevations[indices[j + 1]]) {
+        elevation = elevations[indices[j]];
+      } else {
+        // Both elevations are valid, interpolate.
+        elevation = Math.round((1 - gamma) * elevations[indices[j]] + gamma * elevations[indices[j + 1]]);
+      }
+    }
+
+    // Value
+    double value = (1 - gamma) * ((Number) values[indices[j]]).doubleValue() + gamma * ((Number) values[indices[j + 1]]).doubleValue();
+
+    return new Object[] {tick, location, elevation, value};
   }
 
   @Override
   public String toString() {
     StringBuilder sb = new StringBuilder();
     sb.append(StackUtils.toString(this.percentile));
     sb.append(" ");
+    sb.append("type" + type);
```

Comment:
This will lead to invalid WarpScript, the type needs to be enclosed in single quotes.